### PR TITLE
Bar 49 add an option to switch between federated strategies and try flwr integrated strategies

### PR DESCRIPTION
**Pull request by alex-medvedev-msc**
1. Model and checkpointer mixin (MCMixin)
2. FedAvg, QFedAvg, FedAdam, FedAdagrad wrappers
3. Infer node_count in config from environment

### CODE REVIEW

File: src/fl/server.py [hunk 1]
```diff
@@ -1,12 +1,49 @@
...
+        checkpoint_dir (str): Directory for saving model checkpoints
+
+    Raises:
+        ValueError: If strategy_params.name is unknown
+
+    Returns:
+        FedAvg: flwr Strategy with checkpointing and mlflow logging capabilities
+    """    
+    default_args = OmegaConf.create({
+        "fraction_fit": 0.99,
```
Comment (dkolobok) [inline src/fl/server.py:26]: what is fraction_fit and fraction_eval? a minimum fraction of clients available to fit and evaluate?

**Review by dkolobok — CHANGES_REQUESTED** [review]

File: src/fl/federation/strategy.py [hunk 2]
```diff
@@ -42,45 +42,116 @@ def _calculate_agg_metric(self, metric_name: str, results: RESULTS) -> float:
...
         return sum(losses) / sum(examples)
 
-    def aggregate_evaluate(
-        self,
-        rnd: int,
-        results: RESULTS,
-        failures: List[BaseException],
-    ) -> Tuple[Optional[float], Dict[str, Scalar]]:
+    def log_losses(self, rnd: int, results: RESULTS) -> float:
+        """Logs val_loss, val and train r^2 to mlflow
```
[the remark this answers is not in the document]
Comment (alex-medvedev-msc) [reply src/fl/federation/strategy.py:46]: train_loss and val_loss are averaged over samples by default. However, they depend on phenotype standard deviation

File: src/fl/server.py (hunk 1 above)
Comment (alex-medvedev-msc) [reply src/fl/server.py:26]: Yes, and in our case we want every client to always be available for fit and evaluate

File: src/fl/federation/client.py [hunk 3]
```diff
@@ -47,32 +47,32 @@ def fit(self, parameters, config):
...
-        r2 = r2_score(y_true.detach().cpu().numpy(), preds.detach().cpu().numpy())
-        return r2
-
-    def calculate_train_val_r2(self, trainer: Trainer) -> float:
-        train_loader, val_loader = self.data_module.predict_dataloader()
+        preds = torch.cat(preds, dim=0).detach().cpu().numpy()
+        y_true = torch.cat([batch[1] for batch in iter(loader)]).detach().cpu().numpy()
         
-        return self.calculate_r2_score(trainer, train_loader), self.calculate_r2_score(trainer, val_loader)
+        mse = mean_squared_error(y_true, preds)
```
Comment (dkolobok) [inline src/fl/federation/client.py:55]: Please don't forget that in the future we will continue binary phenotypes as well so MSE won't be relevant there. Maybe we put the list of metrics in the arguments? Maybe as a dict: metrics_dict={'r2': r2_score, 'MSE': mean_squared_error} and in the body:
metrics_values_dict = {k: v(y_true, preds) for k, v in metrics_dict.items()}

File: src/fl/federation/client.py [hunk 4]
```diff
@@ -47,32 +47,32 @@ def fit(self, parameters, config):
...
 
         val_len = self.data_module.val_len()
-        round = self.model.current_round
-        logging.info(f'round: {round}\ttrain_r2: {train_r2:.4f}\tval_r2: {val_r2:.4f}\tval_loss: {val_loss:.3f}\tval_len: {val_len}')
-        mlflow.log_metric('train_r2', train_r2, round)
-        mlflow.log_metric('val_r2', val_r2, round)
-        mlflow.log_metric('val_loss', val_loss, round)
-        return val_loss, val_len, {"val_loss": val_loss, "train_r2": train_r2, "val_r2": val_r2}
+        epoch = self.model.fl_current_epoch()
+        logging.info(f'round: {self.model.current_round}\ttrain_loss: {train_loss:.4f}\ttrain_r2: {train_r2:.4f}\tval_loss: {val_loss:.3f}\tval_r2: {val_r2:.4f}\tval_len: {val_len}')
```
Comment (dkolobok) [inline src/fl/federation/client.py:72]: let's print epoch here as well?

File: src/fl/federation/client.py [hunk 5]
```diff
@@ -47,32 +47,32 @@ def fit(self, parameters, config):
...
+        mlflow.log_metric('train_loss', train_loss, epoch)
+        mlflow.log_metric('train_r2', train_r2, epoch)
+        mlflow.log_metric('val_r2', val_r2, epoch)
+        mlflow.log_metric('val_loss', val_loss, epoch)
+        logging.info(f'val_loss type: {type(val_loss)}, val_len type: {type(val_len)}')
+        return val_loss, val_len, {"val_loss": val_loss, "train_loss": train_loss, "train_r2": train_r2, "val_r2": val_r2}
```
Comment (dkolobok) [inline src/fl/federation/client.py:78]: this is a quite general question: why do we not inherit this and some other methods from the code for local models? as far as I understand, a FL client class should be very similar to a local model class. One of the main differences is to keep track of the round number in FL client but not in local. But many other things can be inherited I guess

**Review by dkolobok — APPROVED** [review]
I approve, please address the questions and then merge. If we decide on some changes for "a quite general question", let's make them in a different task/branch

File: src/fl/federation/client.py (hunk 4 above)
Comment (alex-medvedev-msc) [reply src/fl/federation/client.py:72]: epoch is always a self.current_round*number_of_epochs_in_round in client evaluate function

**Pull request merged**